File: playground/01-Introductions/binary_search.py
def binary_search (list = [], item = 0):
  low = 0
  high = len(list) - 1

  while low <= high:
    mid = ( low + high ) // 2
    guess = list[mid]

    if guess == item:
      return mid
    if guess > item:
      high = mid - 1
    else:
      low = mid + 1

  return None

ordered_list = [0, 3, 4, 5, 6, 7, 8, 8, 11, 12, 13, 15, 16]

print('Index of 11 is: ', binary_search(ordered_list, 11))

File: playground/01-Introductions/binary-search.ts
const binarySearch = <T>(list: T[], item: T): number | null => {
	let low = 0;
	let high = list.length - 1;

	while (low <= high) {
		let mid = Math.floor((low + high) / 2);
		let guess = list[mid];

		if (item === guess) return mid;
		if (item < guess) {
			high = mid - 1;
		} else {
			low = mid + 1;
		}
	}

	return null;
};

const orderedList = [0, 3, 4, 5, 6, 7, 8, 8, 11, 12, 13, 15, 16];
console.log(binarySearch(orderedList, 8));
